pkg/daemon/cleanup/cleanup.go

```diff
@@ -0,0 +1,99 @@
+limitations under the License.
+*/
+
+package cleanup
+
+import (
+	"context"
+	"net"
+
+	"github.com/pkg/errors"
+	core "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/controller"
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/manager"
+	"github.com/openshift/windows-machine-config-operator/pkg/nodeconfig"
+	"github.com/openshift/windows-machine-config-operator/pkg/servicescm"
+)
+
+// Deconfigure removes an instance from the cluster. If preserveNode is true, the node object is not deleted.
+// If we are able to get the services ConfigMap tied to the desired version, all services defined in it are cleaned up.
+// TODO: Otherwise, perform cleanup based on a combination of the OpenShift managed tag and latest services ConfigMap.
+func Deconfigure(c client.Client, netInterfaceAddresses []net.Addr, svcMgr manager.Manager, preserveNode bool) error {
+	var node *core.Node
+	ctx := context.TODO()
+	cm := &core.ConfigMap{}
+	// Fetch the CM of the desired version
+	err := func() error {
+		var err error
+		node, err = controller.GetAssociatedNode(c, netInterfaceAddresses)
+		if err != nil {
+			return err
+		}
+		desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+		if !present {
+			return errors.Wrapf(err, "node missing desired version annotation")
+		}
+		return c.Get(ctx,
+			client.ObjectKey{Namespace: controller.WMCONamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	}()
+	if err != nil {
+		// TODO: best effort cleanup of all OpenShift managed services and those in the latest services ConfigMap
+		// https://issues.redhat.com/browse/WINC-733
+		return err
+	}
+
+	// Happy path, no issues getting the desired services ConfigMap
+	cmData, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+	if err = removeCMServices(svcMgr, cmData.Services); err != nil {
+		return err
+	}
+
+	// Delete node if needed
+	if preserveNode || node == nil {
+		return nil
+	}
+	klog.Infof("deleting node %s", node.GetName())
+	return c.Delete(ctx, node)
+}
+
+// removeCMServices removes all the services defined in the given ConfigMap from this instance.
+func removeCMServices(svcMgr manager.Manager, services []servicescm.Service) error {
+	// Build up log message and failures
+	servicesRemoved := []string{}
+	failedRemovals := []error{}
+	// The services are ordered by increasing priority already, so stop them in reverse order to avoid dependency issues
+	for i := len(services) - 1; i >= 0; i-- {
+		service := services[i]
+		if err := svcMgr.DeleteService(service.Name); err != nil {
+			failedRemovals = append(failedRemovals, err)
+		} else {
+			servicesRemoved = append(servicesRemoved, service.Name)
+		}
+	}
+	klog.Infof("removed services: %q", servicesRemoved)
+	if len(failedRemovals) > 0 {
+		return errors.Errorf("%#v", failedRemovals)
```

Comment:
What is the output difference with `%v` and `%#v`?

Reply:
Benefit of comma separating the errors:
`%#v`
```
[]error{failed to delete service "test1": <something Windows OS returns>, failed to stop service "test2": <something Windows OS returns>, failed to open service "test3": <something Windows OS returns>}
```

`%v`
```
 [failed to delete service "test1": <something Windows OS returns> failed to stop service "test2": <something Windows OS returns> failed to open service "test3": <something Windows OS returns>]
```